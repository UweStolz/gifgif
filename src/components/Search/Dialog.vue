<template>
  <v-dialog
    v-model="syncedShowDialog"
    :full-image-data="syncedFullImageData"
    :preview-image-data="syncedpreviewImageData"
    :image-id="syncedImageId"
    :width="$vuetify.breakpoint.xsOnly ? '100%' : '60%'"
    @click:outside="resetProps"
  >
    <v-card>
      <v-card-title>
        <v-spacer />
        <v-btn
          icon
          @click="resetProps"
        >
          <v-icon>{{ icons.mdiClose }}</v-icon>
        </v-btn>
      </v-card-title>
      <v-img
        :src="syncedFullImageData"
        :lazy-src="syncedFullImageData"
        :max-height="maxWindowSize"
        contain
        height="100%"
        width="100%"
      />
      <v-card-actions>
        <v-btn
          large
          icon
          @click="saveImage"
        >
          <v-icon id="downloadIcon">
            {{ icons.mdiDownload }}
          </v-icon>
        </v-btn>
        <v-spacer />

        <v-btn
          v-if="rating !== 0"
          large
          icon
          @click="removeGif"
        >
          <v-icon>
            {{ icons.mdiTrashCan }}
          </v-icon>
        </v-btn>
        <v-rating
          v-model="rating"
          length="5"
          :empty-icon="icons.mdiHeartOutline"
          :full-icon="icons.mdiHeart"
          :half-icon="icons.mdiHeartHalfFull"
          hover
          x-large
          color="red"
          background-color="red"
          @input="updateGifRating"
        />
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  Vue, Component, PropSync, Watch,
} from 'vue-property-decorator';
import {
  mdiClose, mdiDownload, mdiHeartOutline, mdiHeart, mdiHeartHalfFull, mdiTrashCan,
} from '@mdi/js';
import { getBlob } from '@/request';
import { saveAs } from 'file-saver';

@Component
export default class Dialog extends Vue {
  @PropSync('showDialog', { type: Boolean, required: true }) syncedShowDialog!: boolean;

  @PropSync('fullImageData', { type: String, required: true }) syncedFullImageData!: string;

  @PropSync('previewImageData', { type: String, required: true }) syncedpreviewImageData!: string;

  @PropSync('imageId', { type: String, required: true }) syncedImageId!: string;

  @Watch('syncedShowDialog')
  async getRating() {
    const gifData = await this.$store.dispatch('getGifData', `ggid-${this.syncedImageId}`);
    if (gifData) { this.rating = gifData.rating || 0; }
  }

  saveImage(): void {
    saveAs(this.syncedFullImageData, `ggid-${this.syncedImageId}`);
  }

  resetProps(): void {
    this.syncedShowDialog = false;
    this.syncedFullImageData = '';
    this.rating = 0;
  }

  // eslint-disable-next-line class-methods-use-this
  async getBlobForGif(imageData: string) {
    const previewGifBlob = await getBlob(imageData);
    return previewGifBlob;
  }

  async updateGifRating(rating: number) {
    if (this.$store.state.fullImageMode) {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.syncedImageId}`,
        image: await this.getBlobForGif(this.syncedFullImageData),
        preview: await this.getBlobForGif(this.syncedpreviewImageData),
      });
    } else {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.syncedImageId}`,
        image: this.syncedFullImageData,
        preview: this.syncedpreviewImageData,
      });
    }
    const currentCount = await this.$store.dispatch('getGifCount');
    if (this.gifCount !== currentCount) { this.$store.commit('setGifCount', currentCount); }
  }


  async removeGif() {
    await this.$store.dispatch('removeGifData', `ggid-${this.syncedImageId}`);
    this.rating = 0;
    this.gifCount -= 1;
    this.$store.commit('setGifCount', this.$store.state.gifCount - 1);
  }

  icons = {
    mdiClose,
    mdiDownload,
    mdiHeartOutline,
    mdiHeart,
    mdiHeartHalfFull,
    mdiTrashCan,
  }

  gifCount: number = 0;

  rating: number = 0;

  maxWindowSize: number = window.innerHeight / 1.5;
}
</script>

<style scoped>
#trashIcon:hover {
  color: red;
}
#downloadIcon:hover {
  color: green;
}
</style>
