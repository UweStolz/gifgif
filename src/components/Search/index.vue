<template>
  <v-container
    fluid
  >
    <v-row
      align="start"
      justify="center"
    >
      <v-card
        width="50%"
        flat
        tile
      >
        <v-autocomplete
          v-model="selectedTerms"
          :search-input.sync="searchQuery"
          :loading="isLoading"
          :prepend-icon="icons.mdiImageSearch"
          :items="trendingTerms"
          label="GIF search"
          placeholder="Enter or select search tags"
          cache-items
          autofocus
          hide-no-data
          deletable-chips
          chips
          multiple
          clearable
        />
        <v-card-text>{{ searchResults }}</v-card-text>
      </v-card>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Vue, Component, Watch } from 'vue-property-decorator';
import { mdiImageSearch } from '@mdi/js';
import { getSearchGifsFromGiphy, getTrendingSearchTermsFromTenor } from '@/request';

@Component
export default class Search extends Vue {
  @Watch('searchQuery')
  async searchWithQuery(query: string) {
    await this.getSearchResults(query);
  }

    @Watch('selectedTerms')
  async searchWithTerms(term: string) {
    await this.getSearchResults(term);
  }

    async getSearchResults(query: string) {
      this.isLoading = true;
      const response = await getSearchGifsFromGiphy(query);
      this.searchResults = response;
      this.isLoading = false;
    }

    async mounted() {
      this.trendingTerms = await getTrendingSearchTermsFromTenor();
    }

  icons = {
    mdiImageSearch,
  }

  trendingTerms: string[] = [];

  searchResults: null|any = null;

  searchQuery: null|string = null;

  isLoading: boolean = false;

  selectedTerms: null|any = null;
}
</script>
