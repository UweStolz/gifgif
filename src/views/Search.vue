<template>
  <SearchView />
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import SearchView from '../components/Search/index.vue';

 @Component({
   components: {
     SearchView,
   },
 })
export default class Search extends Vue {}
</script>
