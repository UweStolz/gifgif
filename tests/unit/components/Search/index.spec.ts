import Search from '@/components/Search/index.vue';
import shallow, { fetchMock } from '../../../helper';

const mockResponse = {
  results: ['tag1', 'tag2', 'tag3'],
};

describe('Search.vue', () => {
  it('renders properly', () => {
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Search);
    expect(wrapper.element).toMatchSnapshot();
  });
});
