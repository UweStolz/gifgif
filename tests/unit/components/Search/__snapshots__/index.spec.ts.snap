// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue renders properly 1`] = `
<v-container-stub
  fluid="true"
  tag="div"
>
  <dialog-stub
    fullimagedata=""
    imageid=""
    previewimagedata=""
  />
   
  <v-row-stub
    align="start"
    justify="center"
    tag="div"
  >
    <v-card-stub
      flat="true"
      loaderheight="4"
      tag="div"
      tile="true"
      width="80%"
    >
      <v-slide-y-transition-stub
        mode="out-in"
        origin="top center 0"
      >
        <v-card-stub
          flat="true"
          id="s-filterCard"
          loaderheight="4"
          style="display: none;"
          tag="div"
          tile="true"
        >
          <v-overflow-btn-stub
            allowoverflow="true"
            appendicon="$dropdown"
            backgroundcolor=""
            clearable="true"
            clearicon="$clear"
            errorcount="1"
            errormessages=""
            filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
            itemcolor="primary"
            itemdisabled="disabled"
            items="[object Object],[object Object],[object Object],[object Object]"
            itemtext="text"
            itemvalue="value"
            label="Rating"
            loaderheight="2"
            menuprops="[object Object]"
            messages=""
            nodatatext="$vuetify.noDataText"
            rules=""
            successmessages=""
            target="#s-filterCard"
            type="text"
            value=""
            valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
          />
        </v-card-stub>
      </v-slide-y-transition-stub>
       
      <v-combobox-stub
        allowoverflow="true"
        appendicon="$dropdown"
        backgroundcolor=""
        cacheitems="true"
        chips="true"
        clearable="true"
        clearicon="$clear"
        color="red darken-3"
        deletablechips="true"
        delimiters=""
        errorcount="1"
        errormessages=""
        filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
        hidenodata="true"
        hint="A minimum of 3 characters is required"
        itemcolor="primary"
        itemdisabled="disabled"
        items=""
        itemtext="text"
        itemvalue="value"
        label="GIF search"
        loaderheight="2"
        menuprops="[object Object]"
        messages=""
        nodatatext="$vuetify.noDataText"
        persistenthint="true"
        placeholder="Enter or select a search tag"
        prependicon="M15.5,2C18,2 20,4 20,6.5C20,7.38 19.75,8.21 19.31,8.9L22.39,12L21,13.39L17.88,10.32C17.19,10.75 16.37,11 15.5,11C13,11 11,9 11,6.5C11,4 13,2 15.5,2M15.5,4A2.5,2.5 0 0,0 13,6.5A2.5,2.5 0 0,0 15.5,9A2.5,2.5 0 0,0 18,6.5A2.5,2.5 0 0,0 15.5,4M7.5,14.5L4,19H18L13.5,13L10,17.5L7.5,14.5M20,20A2,2 0 0,1 18,22H4A2,2 0 0,1 2,20V6A2,2 0 0,1 4,4H9.5C9.18,4.77 9,5.61 9,6.5A6.5,6.5 0 0,0 15.5,13C16.18,13 16.84,12.89 17.46,12.7L20,15.24V20Z"
        returnobject="true"
        rules=""
        searchinput=""
        successmessages=""
        type="text"
        valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
      />
       
      <v-divider-stub />
       
      <v-expand-transition-stub
        mode="in-out"
      >
        <!---->
      </v-expand-transition-stub>
    </v-card-stub>
  </v-row-stub>
</v-container-stub>
`;
